From fe0063833f355f8f663e47fdfb6e33800537e577 Mon Sep 17 00:00:00 2001
From: Ondrej Jirman
Date: Sun, 26 Jan 2020 00:37:31 +0100
Subject: arm64: dts: pinephone: Add support for the front camera (gc2145)

A few changes are packed into this commit:

- there's no need for a pullup on the I2C gpio lines, there are 2kOhm
  resistors present externally
- we should use open drain driving on the reset/standby signals,
  there are weak pullups externally
- pinctrl for mclk has to be on the csi node, to avoid conflicting
  settings
- pio must not enable PE power supply uncoditionally, it needs to be
  enabled only when one of the cameras are enabled

Signed-off-by: Ondrej Jirman <megous@megous.com>
---
 .../boot/dts/allwinner/sun50i-a64-pinephone.dts    | 101 ++++++++++++++-------
 1 file changed, 68 insertions(+), 33 deletions(-)

diff --git a/arch/arm64/boot/dts/allwinner/sun50i-a64-pinephone.dts b/arch/arm64/boot/dts/allwinner/sun50i-a64-pinephone.dts
index e43f4a3ff5de..20b7dbbfcd51 100644
--- a/arch/arm64/boot/dts/allwinner/sun50i-a64-pinephone.dts
+++ b/arch/arm64/boot/dts/allwinner/sun50i-a64-pinephone.dts
@@ -34,40 +34,13 @@
 		stdout-path = "serial0:115200n8";
 	};
 
-	i2c-csi {
+	i2c_csi: i2c-csi {
 		compatible = "i2c-gpio";
-		sda-gpios = <&pio 4 13 (GPIO_ACTIVE_HIGH|GPIO_OPEN_DRAIN|GPIO_PULL_UP)>; /* PE13 */
-		scl-gpios = <&pio 4 12 (GPIO_ACTIVE_HIGH|GPIO_OPEN_DRAIN|GPIO_PULL_UP)>; /* PE12 */
-		i2c-gpio,delay-us = <5>;
+		sda-gpios = <&pio 4 13 (GPIO_ACTIVE_HIGH | GPIO_OPEN_DRAIN)>; /* PE13 */
+		scl-gpios = <&pio 4 12 (GPIO_ACTIVE_HIGH | GPIO_OPEN_DRAIN)>; /* PE12 */
+		i2c-gpio,delay-us = <1>;
 		#address-cells = <1>;
 		#size-cells = <0>;
-
-		/* Rear camera */
-		ov5640: camera@3c {
-			compatible = "ovti,ov5640";
-			reg = <0x3c>;
-			pinctrl-names = "default";
-			pinctrl-0 = <&csi_mclk_pin>;
-			clocks = <&ccu CLK_CSI_MCLK>;
-			clock-names = "xclk";
-
-			AVDD-supply = <&reg_dldo3>;
-			DOVDD-supply = <&reg_aldo1>;
-			DVDD-supply = <&reg_eldo3>;
-			reset-gpios = <&pio 3 3 GPIO_ACTIVE_LOW>; /* PD3 */
-			powerdown-gpios = <&pio 2 0 GPIO_ACTIVE_HIGH>; /* PC0 */
-
-			port {
-				ov5640_ep: endpoint {
-					remote-endpoint = <&csi_ep>;
-					bus-width = <8>;
-					hsync-active = <1>; /* Active high */
-					vsync-active = <0>; /* Active low */
-					data-active = <1>;  /* Active high */
-					pclk-sample = <1>;  /* Rising */
-				};
-			};
-		};
 	};
 
 	leds {
@@ -171,13 +144,15 @@
 };
 
 &csi {
+	pinctrl-0 = <&csi_pins>, <&csi_mclk_pin>;
 	status = "okay";
 
 	port {
 		#address-cells = <1>;
 		#size-cells = <0>;
 
-		csi_ep: endpoint {
+		csi_ov5640_ep: endpoint@0 {
+			reg = <0>;
 			remote-endpoint = <&ov5640_ep>;
 			bus-width = <8>;
 			hsync-active = <1>; /* Active high */
@@ -185,6 +160,16 @@
 			data-active = <1>;  /* Active high */
 			pclk-sample = <1>;  /* Rising */
 		};
+
+		csi_gc2145_ep: endpoint@1 {
+			reg = <1>;
+			remote-endpoint = <&gc2145_ep>;
+			bus-width = <8>;
+			hsync-active = <1>;
+			vsync-active = <1>;
+			data-active = <1>;
+			pclk-sample = <1>;
+		};
 	};
 };
 
@@ -226,6 +211,56 @@
 	status = "okay";
 };
 
+&i2c_csi {
+	gc2145: front-camera@3c {
+		compatible = "galaxycore,gc2145";
+		reg = <0x3c>;
+		clocks = <&ccu CLK_CSI_MCLK>;
+		clock-names = "xclk";
+
+		AVDD-supply = <&reg_dldo3>;
+		DVDD-supply = <&reg_aldo1>;
+		IOVDD-supply = <&reg_eldo3>;
+		reset-gpios = <&pio 4 16 (GPIO_ACTIVE_LOW | GPIO_OPEN_DRAIN)>; /* PE16 */
+		enable-gpios = <&pio 4 17 (GPIO_ACTIVE_LOW | GPIO_OPEN_DRAIN)>; /* PE17 */
+
+		port {
+			gc2145_ep: endpoint {
+				remote-endpoint = <&csi_gc2145_ep>;
+				bus-width = <8>;
+				hsync-active = <1>;
+				vsync-active = <1>;
+				data-active = <1>;
+				pclk-sample = <1>;
+			};
+		};
+	};
+
+	ov5640: rear-camera@4c {
+		compatible = "ovti,ov5640";
+		reg = <0x4c>;
+		clocks = <&ccu CLK_CSI_MCLK>;
+		clock-names = "xclk";
+
+		AVDD-supply = <&reg_dldo3>;
+		DOVDD-supply = <&reg_aldo1>; /* shared with AFVCC */
+		DVDD-supply = <&reg_eldo3>;
+		reset-gpios = <&pio 3 3 (GPIO_ACTIVE_LOW | GPIO_OPEN_DRAIN)>; /* PD3 */
+		powerdown-gpios = <&pio 2 0 (GPIO_ACTIVE_HIGH | GPIO_OPEN_DRAIN)>; /* PC0 */
+
+		port {
+			ov5640_ep: endpoint {
+				remote-endpoint = <&csi_ov5640_ep>;
+				bus-width = <8>;
+				hsync-active = <1>; /* Active high */
+				vsync-active = <0>; /* Active low */
+				data-active = <1>;  /* Active high */
+				pclk-sample = <1>;  /* Rising */
+			};
+		};
+	};
+};
+
 &i2c0 {
 	pinctrl-names = "default";
 	pinctrl-0 = <&i2c0_pins>;
@@ -351,7 +386,7 @@
 };
 
 &pio {
-	vcc-pe-supply = <&reg_aldo1>;
+	//vcc-pe-supply = <&reg_aldo1>;
 	vcc-pg-supply = <&reg_dldo4>;
 
 	i2s0_pins: i2s0-pins {
-- 
cgit v1.2.2

 
